### PR TITLE
[MAINTENANCE] Output Consistent Data Format from "table.head" Metric for every ExecutionEngine

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
Standardize output of `table.head` metric on `pd.DataFrame`, thereby removing the need to predicate on `ExecutionEngine` in `MetricsCalculator` and other callers.  Consequently, `Batch.head()` in New Datasources becomes "a one-liner proxy" to `MetricsCalculator.head()` -- and similar simplifications have been applied to a few other places.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1654/GREAT-1637
- 
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/experimental/datasources/interfaces.py [hunk 1]
```diff
@@ -505,7 +507,7 @@ def update_forward_refs(cls):
     @validate_arguments
     def head(
         self,
-        n_rows: Optional[StrictInt] = None,
```
Comment (NathanFarmer) [inline great_expectations/experimental/datasources/interfaces.py:508]: 🙇🏻

File: great_expectations/expectations/metrics/table_metrics/table_head.py [hunk 2]
```diff
@@ -199,11 +200,11 @@ def _spark(
         metric_value_kwargs: dict,
         metrics: dict[str, Any],
         runtime_configuration: dict,
-    ) -> list[pyspark_sql_Row] | pyspark_sql_Row:
```
Comment (NathanFarmer) [inline great_expectations/expectations/metrics/table_metrics/table_head.py:202]: @alexsherstinsky I'm surprised this passes tests in `test_metric_configuration.py` and `test_sparkdf_execution_engine.py` considering how much we use this metric there.

**Review by NathanFarmer — COMMENTED** [review]
Can you run the full test suite on this PR please?

**Comment by alexsherstinsky** [comment]
> Can you run the full test suite on this PR please?

I have and it looked like it passed them all.  It is running them again, since I updated with the latest from develop.  I did go through the entire code base and carefully updated each use.  Are you seeing something that does not look right?  Thanks, @NathanFarmer !

**Comment by NathanFarmer** [comment]
> > Can you run the full test suite on this PR please?
> 
> I have and it looked like it passed them all. It is running them again, since I updated with the latest from develop. I did go through the entire code base and carefully updated each use. Are you seeing something that does not look right? Thanks, @NathanFarmer !

@alexsherstinsky This should be expected to fail with your change I would think: [https://github.com/great-expectations/great_expectations/blob/10f884135fe27e03c991[…]87ff353/tests/execution_engine/test_sparkdf_execution_engine.py](https://github.com/great-expectations/great_expectations/blob/10f884135fe27e03c9919b555573bc4b687ff353/tests/execution_engine/test_sparkdf_execution_engine.py#L108)

**Comment by alexsherstinsky** [comment]
> https://github.com/great-expectations/great_expectations/blob/10f884135fe27e03c991[…]87ff353/tests/execution_engine/test_sparkdf_execution_engine.py

@NathanFarmer This passes, because the assertion looks at the `Batch` data (a level lower than the output of the metric).  Thanks.

**Comment by alexsherstinsky** [comment]
@NathanFarmer Here is how it runs for me locally:
```
pytest --mysql --mssql --postgresql --spark --cache-clear --full-trace tests/execution_engine/test_sparkdf_execution_engine.py::test_reader_fn_parameters -svv

PASSED

====================================================================================================================== warnings summary ======================================================================================================================
venv/lib/python3.9/site-packages/botocore/httpsession.py:18
  /Users/alexsherstinsky/Development/GreatExpectations/great_expectations/venv/lib/python3.9/site-packages/botocore/httpsession.py:18: DeprecationWarning: 'urllib3.contrib.pyopenssl' module is deprecated and will be removed in a future release of urllib3 2.x. Read more in this issue: https://github.com/urllib3/urllib3/issues/2680
    from urllib3.contrib.pyopenssl import orig_util_SSLContext as SSLContext

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================================ 1 passed, 1 warning in 9.80s ================================================================================================================
```

**Review by NathanFarmer — APPROVED** [review]

File: great_expectations/validator/metrics_calculator.py [hunk 3]
```diff
@@ -1,13 +1,7 @@
...
 from typing import Any, Dict, List, Optional, Set, Tuple, Union
 
-import great_expectations.exceptions as gx_exceptions
-from great_expectations.execution_engine import (
-    ExecutionEngine,
-    PandasExecutionEngine,
-    SparkDFExecutionEngine,
-    SqlAlchemyExecutionEngine,
-)
+from great_expectations.execution_engine import ExecutionEngine  # noqa: TCH001
```
Comment (Kilo59) [inline great_expectations/validator/metrics_calculator.py:4]: Should move this to the `TYPE_CHECKING` block.

Comment (alexsherstinsky) [reply great_expectations/validator/metrics_calculator.py:4]: @Kilo59 That did not work in this PR, despite what RUFF said.  Hence, I had to escape it.  However, it seems to be working fine in #7094 -- I do not understand the reason for this inconsistency.

**Pull request merged**